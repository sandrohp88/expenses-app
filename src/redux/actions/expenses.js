import expensesRef from '../../firebase/firebase'
// ADD_EXPENSE
const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense
})

// ADD_EXPENSE_ASYNC
const addExpenseAsync = (expenseData = {}) => {
  return async dispatch => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData
    const expense = { description, note, amount, createdAt }
    const response = await expensesRef.add(expense)
    return dispatch(addExpense({ id: response.id, ...expense }))
  }
}

// REMOVE_EXPENSE
const removeExpense = (id = '') => ({ type: 'REMOVE_EXPENSE', id })

// EDIT_EXPENSE
const editExpense = (id, updates) => ({ type: 'EDIT_EXPENSE', id, updates })

// SET_EXPENSES
const setExpenses = expenses => ({ type: 'SET_EXPENSES', expenses })

// GET_EXPENSES_ASYNC
const getExpensesAsync = () => {
  return async dispatch => {
    const response = await expensesRef.get()
    const expenses = []
    response.forEach(expense => {
      const {
        description = '',
        note = '',
        amount = 0,
        createdAt = 0
      } = expense.data()
      expenses.push({id:expense.id, description, note, amount, createdAt })
    })
    return dispatch(setExpenses(expenses))
  }
}

export {
  getExpensesAsync,
  addExpense,
  addExpenseAsync,
  removeExpense,
  editExpense,
  setExpenses
}
