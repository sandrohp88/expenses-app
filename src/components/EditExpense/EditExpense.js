import React from 'react'
import { connect } from 'react-redux'
import { ExpenseForm } from '../ExpenseForm/ExpenseForm'
import { editExpense, removeExpense } from '../../redux/actions/expenses'

export class EditExpense extends React.Component {
  onSubmit = expense => {
    this.props.editExpense(this.props.expense.id, expense)
    this.props.history.push('/')
  }

  onClick = () => {
    this.props.removeExpense(this.props.expense.id)
    this.props.history.push('/')
  }
  render() {
    return (
      <div>
        <ExpenseForm onSubmit={this.onSubmit} expense={this.props.expense} />
        <button onClick={this.onClick}>Remove</button>
      </div>
    )
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      expense => expense.id === props.match.params.id
    )
  }
}
const mapDispatchToProps = dispatch => ({
  removeExpense: id => dispatch(removeExpense(id)),
  editExpense: (id, updates) => dispatch(editExpense(id, updates))
})
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpense)
