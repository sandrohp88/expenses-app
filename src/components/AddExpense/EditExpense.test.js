import React from 'react'
import { shallow } from 'enzyme'
import { EditExpense } from '../../components/EditExpense/EditExpense'
import expenses from '../../sampleData/expenses'

let editExpense, history, wrapper, removeExpense
beforeEach(() => {
  editExpense = jest.fn()
  removeExpense = jest.fn()
  history = { push: jest.fn() }
  wrapper = shallow(
    <EditExpense
      expense={expenses[2]}
      editExpense={editExpense}
      history={history}
      removeExpense={removeExpense}
    />
  )
})

test('render edit expense without crash', () => {
  expect(wrapper).toMatchSnapshot()
})
test('handle editExpense', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2])
  expect(editExpense).toHaveBeenLastCalledWith(expenses[2].id, expenses[2])
  expect(history.push).toHaveBeenLastCalledWith('/')
})
test('handle remove expense', () => {
  wrapper.find('button').simulate('click')
  expect(removeExpense).toHaveBeenLastCalledWith(expenses[2].id)
  expect(history.push).toHaveBeenLastCalledWith('/')
})
