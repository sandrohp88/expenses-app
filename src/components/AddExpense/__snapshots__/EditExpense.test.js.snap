// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render edit expense without crash 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 4500,
        "createdAt": 345600000,
        "description": "Credit Card",
        "id": "3",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
